### PR TITLE
[Train] Refactor Backends

### PR DESCRIPTION
**Pull request by amogkam**
Refactors backends to move them to the top level `ray.train` directory.

- Removes `backends` module
- Removes backend specific configurations from the top-level `ray.train` module
- Removes the "safe check" for backend dependencies. Now, if you import from `ray.train.torch` it is expected that torch is installed.
- Adds `test_callbacks` to CI (this was missed before) and fixes TBX test.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/trainer.py [hunk 1]
```diff
@@ -40,13 +38,28 @@ def noop():
...
+    try:
+        import importlib
+        config_cls = getattr(
+            importlib.import_module(f"ray.train"
+                                    f".{backend_name}"),
+            BACKEND_NAME_TO_CONFIG_CLS_NAME[backend_name])
+        return config_cls
+    except ImportError:
+        raise ValueError(f"{backend_name} is not installed. Please "
+                         f"install {backend_name} to use this backend.")
```
Comment (matthewdeng) [inline python/ray/train/trainer.py:60]: This handles the case if the user passes a `backend_name` string pretty cleanly. Do we need similar wrapping logic if the user directly passes in a `BackendConfig`? Will the error message in that scenario be clear?

Comment (amogkam) [reply python/ray/train/trainer.py]: Good point, fixed!

File: python/ray/train/trainer.py [hunk 2]
```diff
@@ -172,12 +181,11 @@ def _get_backend_config(
...
             return backend
         elif isinstance(backend, str):
-            try:
-                return BACKEND_NAME_TO_CONFIG_CLS[backend]()
-            except KeyError:
+            if backend not in BACKEND_NAME_TO_CONFIG_CLS_NAME:
                 raise ValueError(f"Invalid backend: {backend}. "
                                  f"Supported string values are: "
-                                 f"{BACKEND_NAME_TO_CONFIG_CLS.keys()}")
+                                 f"{BACKEND_NAME_TO_CONFIG_CLS_NAME.keys()}")
```
Comment (matthewdeng) [inline python/ray/train/trainer.py:187]: nit: Move this into `get_backend_config_cls` - keeps all logic around `BACKEND_NAME_TO_CONFIG_CLS_NAME` in 1 method.

File: doc/source/train/api.rst [hunk 3]
```diff
@@ -30,21 +30,21 @@ Backend Configurations
...
-.. autoclass:: ray.train.TensorflowConfig
+.. autoclass:: ray.train.tensorflow.TensorflowConfig
 
 .. _train-api-horovod-config:
 
 HorovodConfig
 ~~~~~~~~~~~~~
 
-.. autoclass:: ray.train.HorovodConfig
+.. autoclass:: ray.train.horovod.HorovodConfig
```
Comment (matthewdeng) [inline doc/source/train/api.rst:47]: Lint is failing for this 😞

**Review by matthewdeng — APPROVED** [review]
LGTM besides the lint issue, thanks for cleaning this up!

File: python/ray/train/trainer.py (hunk 2 above)
Comment (amogkam) [reply python/ray/train/trainer.py]: Done!

**Pull request merged**